### DISCUSSION TITLE
Support multi-sized (e.g. pyramid) TIFFs

### DISCUSSION THREAD

**Discussion by Christoph-Wagner** [SixLabors/ImageSharp#1982] [Ideas]
Currently, TIFFs with multiple frame sizes are not supported:

```C#
foreach (ImageFrame<TPixel> frame in frames)
{
    if (frame.Size() != root.Size())
    {
        TiffThrowHelper.ThrowNotSupported("Images with different sizes are not supported");
    }
}
```

There is a TODO comment: `// TODO: Tiff frames can have different sizes` but @JimBobSquarePants asked me to create a feature request as reminder ;)

**Comment by antonfirsov** [comment]
ImageSharp has a core architectural assumption that an image has a single size (even when it consists of multiple frames), so this won't be easy to achieve.

We can pick the biggest frame, and extend the rest, but I think this is not what users want, since it would loose information.

The only solution I can think of is a custom, TIFF-specific API, that would return an `Image<T>[]` containing a separate `Image<T>` for each frame. Any better ideas?

**Reply by Christoph-Wagner** [reply]
For my personal use case (seeing if I can replace our current *very old* solution with something ImageSharp based), it might be enough to get the largest frame (I’d have to see what the performance penalty for resizing would be; obviously a tiff-specific API would make things even easier, but it’s also ugly), but @JimBobSquarePants [wrote](https://old.reddit.com/r/csharp/comments/sn8j15/six_labors_announcing_imagesharp_200/hw25njz/) on reddit:

> We currently constrain multi frame images to the same dimensions though we pad out gif frames on decode. I’d like to figure out a way to allow different sized frames.

which I’d guess would be the more *interesting* solution ;)

**Reply by JimBobSquarePants** [reply]
@antonfirsov  I wanna have a serious look at this for V3. If we then deem it a no-go, then so be it.

One thing that put me off before was tracking things like Gif Fame offset when transforming an image frames. We could potentially leave Gif as-is and continue to pad but for formats that don't offset frames it might be doable.

**Reply by brianpopow** [reply]
This is not an TIFF exclusive issue. For example animated webp images also contain images with different sizes.

**Reply by JimBobSquarePants** [reply]
They'll work in the same way as Gif though won't they where an offset can be read to allow padding? Compression is good for empty spaces so padding by the offset on decode was a simple way to deal with the issue. 

I wonder.... Do we just pad out the Tiff files? Sure it's more expensive memory wise but it's a LOT easier for us.

**Reply by brianpopow** [reply]
> They'll work in the same way as Gif though won't they where an offset can be read to allow padding? Compression is good for empty spaces so padding by the offset on decode was a simple way to deal with the issue.

Yes very similar to gif.

> I wonder.... Do we just pad out the Tiff files? Sure it's more expensive memory wise but it's a LOT easier for us.

We currently throw an `ThrowNotSupported` if the size does not match the root frame

**Reply by JimBobSquarePants** [reply]
Sorry, that "Do we" was meant as a question for potential future behavior. Does each frame contain offset info? (_I've pretty much forgotten everything I read on Tiff - Well tried to anyway_) If so and if the Tiff container header reports the size as matching the largest (_not necessarily root_) frame then we can be confident that we can pad each frame correctly to make them all the same size.

**Reply by antonfirsov** [reply]
We can also add the original size to `TiffFrameMetadata`, so users can crop out the original region from the padded frame if that's what they need.

As a separate 3.0 feature, I think it would be still desirable to enable *streamed* processing of multi-frame images, regardless of the format. For tiff would enable processing frames using their original size. It would also enable optimizations for all the other formats. When someone opens a 100 frame gif today, it can eat up all their memory.

**Reply by JimBobSquarePants** [reply]
>I think it would be still desirable to enable streamed processing of multi-frame images, regardless of the format. 

Yep, absolutely worth a look at.

**Reply by brianpopow** [reply]
I dont think the tiff format provides offset info as gif or webp and the first one needs to be the largest, but i am not 100% sure about that.

@Christoph-Wagner could you provide an example image, so i could look for that info there?

**Reply by tocsoft** [reply]
we could always just remove the frame size limits and instead move the validation down into the gif encoder to validate all frames are the same size...  it should be a simple one liner to force an image, that happens to have multiple differently size frames, to have all frames to the same size by just be a calling resize with the first frames size and having it no-op if the frame sizes match the target size.

**Reply by Christoph-Wagner** [reply]
@brianpopow
[SKC1H3.zip](https://github.com/SixLabors/ImageSharp/files/8023617/SKC1H3.zip)

Not sure if it’s relevant, but the image is created with the following Magick.NET (ImageMagick wrapper) options:

```C#
image.Format = MagickFormat.Ptif;
image.SetAttribute("define", "tiff:tile-geometry=256x256 tiff:ignore-layers");
 ```

**Reply by brianpopow** [reply]
I have checked the example image: tiff format does not provide offset info as gif or animated webp. The example is a typical mipmap image/texture: The first image is the biggest, the second is half the size and so on. I could not find any specifics in the spec, that the first has to be the largest (probably will be, but I guess its not for sure). 

One simple solution would be to make all images the same size as the biggest and decode smaller onces at position (0, 0), while leaving the rest of the image as is. We need to decide then howto treat the undefined pixels (should they be black, white, transparent, etc.)

Until we decide howto deal with images of different sizes, a easy solution for this would be to introduce a TiffDecoderOption similar to the GifDecoderOption: `FrameDecodingMode`, allowing to decode the first frame only. Then the user could manually resize the other mipmap textures from the first one.

**Reply by JimBobSquarePants** [reply]
That sounds like a plan  to me @brianpopow Nice!

**Reply by brianpopow** [reply]
With #2013 merged to main, the Tiff Decoder now has an option `FrameDecodingMode` to only decode the first frame or all.

**Reply by JimBobSquarePants** [reply]
I've made a start on supporting multi-size Tiff files with https://github.com/SixLabors/ImageSharp/pull/2788

Decoding works well, we simply pin the pixel data to the top/left of a frame matching the root frame dimensions.

 I just need to put some thought to handling efficient re-encoding so that we do not encode all that dead space.

**Comment by JimBobSquarePants** [comment]
According to the following information Tiff supports X/Y position tags for the image.
https://awaresystems.be/imaging/tiff/tifftags/search.html?q=position&Submit=Find+Tags

<table cellspacing="0" cellpadding="4" border="1">
<tbody><tr><th colspan="2">Code</th><th rowspan="2">Name</th><th rowspan="2">Short description</th></tr>
<tr><th>Dec</th><th>Hex</th></tr>
</tbody><tbody align="left" valign="top">
<tr><td>286</td><td>0x011E</td><td nowrap=""><a href="xposition.html">XPosition</a></td><td>X position of the image.</td></tr>
<tr><td>287</td><td>0x011F</td><td nowrap=""><a href="yposition.html">YPosition</a></td><td>Y position of the image.</td></tr>
</tbody>
</table>

According to the TIFF 6 spec, they are in resolution units. so to work out the pixel position they would be multiplied by X/YResolution and rounded to the nearest int.

I have found nothing to confirm this but I don't think any frames that follow the first frame can be larger than the first frame so we can always take that frame as the image size. We can check X/Y Position for subsequent frames that are smaller and offset them in the buffer like we do during Gif decode. 

When encoding we can either allow setting a background color and trimming that or simply trim transparent like I do with Gif in https://github.com/SixLabors/ImageSharp/pull/2455 and set those tags appropriately.

**Reply by boukenka** [reply]
Any update on this proposition ?

**Reply by JimBobSquarePants** [reply]
PRs are welcome. If there was an update it would be documented.

**Reply by JimBobSquarePants** [reply]
Turns out that those tags are for the X/Y positions on a printed page. We don't need them for decoding.

**Reply by JimBobSquarePants** [reply]
PR open #2788
